daemon/cmd/hubble.go

```diff
@@ -158,3 +165,100 @@ func (d *Daemon) launchHubble() {
+func (d *Daemon) GetIdentity(securityIdentity uint32) (*models.Identity, error) {
+	ident := d.identityAllocator.LookupIdentityByID(context.Background(), identity.NumericIdentity(securityIdentity))
+	if ident == nil {
+		return nil, fmt.Errorf("identity %d not found", securityIdentity)
+	}
+	return identitymodel.CreateModel(ident), nil
+}
+
+// GetEndpointInfo returns endpoint info for a given IP address. Hubble uses this function to populate
+// fields like namespace and pod name for local endpoints.
+//
+//  - EndpointGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L34
+func (d *Daemon) GetEndpointInfo(ip net.IP) (endpoint v1.EndpointInfo, ok bool) {
+	ep := d.endpointManager.LookupIP(ip)
+	if ep == nil {
+		return nil, false
+	}
+	return ep, true
+}
+
+// GetNamesOf implements DNSGetter.GetNamesOf. It looks up DNS names of a given IP from the
+// FQDN cache of an endpoint specified by sourceEpID.
+//
+//  - DNSGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L27
+func (d *Daemon) GetNamesOf(sourceEpID uint32, ip net.IP) []string {
+	ep := d.endpointManager.LookupCiliumID(uint16(sourceEpID))
+	if ep == nil {
+		return nil
+	}
+	return ep.DNSHistory.LookupIP(ip)
+}
+
+// GetServiceByAddr looks up service by IP/port. Hubble uses this function to annotate flows
+// with service information.
+//
+//  - ServiceGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L52
+func (d *Daemon) GetServiceByAddr(ip net.IP, port uint16) (flowpb.Service, bool) {
+	addr := loadbalancer.L3n4Addr{
+		IP: ip,
+		L4Addr: loadbalancer.L4Addr{
+			Port: port,
+		},
+	}
+	namespace, name, ok := d.svc.GetServiceNameByAddr(addr)
+	if !ok {
+		return flowpb.Service{}, false
+	}
+	return flowpb.Service{
+		Namespace: namespace,
+		Name:      name,
+	}, true
+}
+
+// GetK8sMetadata returns the Kubernetes metadata for the given IP address.
+// It implements hubble parser's IPGetter.GetK8sMetadata.
+func (d *Daemon) GetK8sMetadata(ip net.IP) *ipcache.K8sMetadata {
+	if ip == nil {
+		return nil
+	}
+	return ipcache.IPIdentityCache.GetK8sMetadata(ip.String())
+}
+
+// LookupSecIDByIP returns the security ID for the given IP. If the security ID
+// cannot be found, ok is false.
+// It implements hubble parser's IPGetter.LookupSecIDByIP.
+func (d *Daemon) LookupSecIDByIP(ip net.IP) (id ipcache.Identity, ok bool) {
+	if ip == nil {
+		return ipcache.Identity{}, false
+	}
+
+	if id, ok = ipcache.IPIdentityCache.LookupByIP(ip.String()); ok {
+		return id, ok
+	}
+
+	ipv6Prefixes, ipv4Prefixes := d.GetCIDRPrefixLengths()
+	prefixes := ipv4Prefixes
+	bits := net.IPv4len * 8
+	if ip.To4() == nil {
+		prefixes = ipv6Prefixes
+		bits = net.IPv6len * 8
+	}
+	for _, prefixLen := range prefixes {
+		cidr := net.IPNet{
```

Comment:
nit: You can optimize this further by check for `prefixLen == bits` and skip the lookup in that case (since that is a full IP lookup that you already did).